### ISSUE TITLE
Cannot instantiate `Transformers("microsoft/Phi-3-mini-4k-instruct")`

### ISSUE THREAD

**Issue by hudson-ai** [guidance-ai/guidance#958]
**The bug**

The issue seems to be with [this line](https://github.com/guidance-ai/guidance/blob/be2be221ee6ee7cda3606ddd6f93544b947507c6/guidance/models/transformers/_transformers.py#L104). The `'<0x20>'` token is not surviving tokenize/detokenize round-trip, and we get an `IndexError`.

**To Reproduce**

```python
from guidance import models
model = models.Transformers("microsoft/Phi-3-mini-4k-instruct")
```

> File ~/pkg/guidance-ai/guidance/models/transformers/_transformers.py:104, in TransformersTokenizer.__init__(self, model, transformers_tokenizer, chat_template, ignore_bos_token, **kwargs)
    102 if hasattr(transformers_tokenizer, "convert_tokens_to_string"):
    103     token_str = transformers_tokenizer.convert_tokens_to_string([token])
--> 104     roundtrip_id = transformers_tokenizer.encode(token_str)[0]
    105     if roundtrip_id == i:
    106         byte_coded = token_str.encode()
IndexError: list index out of range

**System info (please complete the following information):**
 - OS (e.g. Ubuntu, Windows 11, Mac OS, etc.): Mac OS
 - Guidance Version (`guidance.__version__`): current head

@riedgar-ms @ambisinister tag :)

**Comment by Harsha-Nori** [comment]
<img width="953" alt="image" src="https://github.com/user-attachments/assets/f6024391-69dd-48cd-92cd-4f9e77425201">

**Comment by hudson-ai** [comment]
Nothing makes sense and I can't repro on my other machine...

**Comment by ambisinister** [comment]
I'll take a look in a minute also

**Comment by ambisinister** [comment]
Can't repro on my macbook with the following

```
import guidance
from guidance import gen, models

model = models.Transformers("microsoft/Phi-3-mini-4k-instruct", echo=False)

@guidance
def gen_test(lm):
    lm = lm + "Scott is a " + gen(name="test", max_tokens=40) +"-Hi there"
    return lm

model += gen_test()

print(str(model))
```

output
```
Scott is a doctor who sees 5 adult patients and 4 child patients every hour. Adult visits cost $50, while child visits cost $30. If Scott works for 8 hours-Hi there
```

**Comment by hudson-ai** [comment]
Ok, thanks for the support y'all :)
Clearly something funky going on on my side, maybe something stuck in a cache from my PR branch...

**Comment by hudson-ai** [comment]
@nking-1 can repro, so at least I'm not crazy!

**Comment by Harsha-Nori** [comment]
So weird...my best guess is that there may be old versions of phi-3 (which has had updates recently) in the HF cache on your local machines? I downloaded a fresh copy when I tried on my side.

**Comment by hudson-ai** [comment]
Ooh very funky. @Harsha-Nori it looks like it comes down to the difference between the `use_fast=True` and `use_fast=False` paths taken [when instantiating the `AutoTokenizer`](https://github.com/guidance-ai/guidance/blob/181ea1ec5875ca3cd81ea58717b37102f50ef28d/guidance/models/transformers/_transformers.py#L203C1-L217C53)...

I can repro by uninstalling `sentencepiece`, which causes the `use_fast=False` branch to fail (silently, I may add... that exception handler should probably be tightened...). Can you confirm on your side?

I'm still totally unsure why this is causing the llguidance PR branch to fail CI... The workflow still appears to be installing sentencepiece...

**Comment by hudson-ai** [comment]
Tricky tricky. PR removed protobuf dependency, causing a silent exception in the exact same place:
```
ImportError: 
The new behaviour of LlamaTokenizer (with `self.legacy = False`) requires the protobuf library but it was not found in your environment. Checkout the instructions on the
installation page of its repo: https://github.com/protocolbuffers/protobuf/tree/master/python#installation and follow the ones
```

**Comment by riedgar-ms** [comment]
Sorry, only just saw this. I have been wondering if we should be tweaking the `TransformersTokenizer` slightly. Specifically swapping these two blocks:

https://github.com/guidance-ai/guidance/blob/181ea1ec5875ca3cd81ea58717b37102f50ef28d/guidance/models/transformers/_transformers.py#L76
https://github.com/guidance-ai/guidance/blob/181ea1ec5875ca3cd81ea58717b37102f50ef28d/guidance/models/transformers/_transformers.py#L89

Basically, prioritise @ambisinister 's method of constructing the byte decoder over the one from sentencepiece (if both are available).

**Comment by hudson-ai** [comment]
@riedgar-ms it looks like the `get_vocab` approach fails when the `use_fast=False` path is taken (at least for `microsoft/Phi-3-mini-4k-instruct`), and that error was obfuscated by the fact that the sentencepiece is prioritized. It may be worth looking into that issue before prioritizing the new implementation.

There's another issue lurking here though -- the silent fallback to a different tokenizer made this a really difficult bug to track down/repro. I'd advocate for separating the except block into a few cases...

1. `ImportError` -- transformers will raise this if the user needs to install an additional dependency for the `fast=False` path to work. I think we should **at a minimum** give a warning if we are going to fall-back in this case, but I'd actually maybe argue for just raising the exception.
2. `AssertionError` -- tokenizer is missing some bytes. Falling back is fine, but maybe it's a warning?
3. Any other error -- maybe don't catch at all until people raise new issues...?

My biggest caution here is that I don't actually know *why* we're falling back here. Was the original intention to fall back in just case 1 and 2? Just 2 and 3? etc.

**Comment by Saibo-creator** [comment]
I encountered the same error on linux, and it turned out that `sentencepiece` wasn't installed. Installing `sentencepiece` resolved the problem.
